backend/icpy/services/remote_fs_adapter.py

```
@@ -281,6 +396,202 @@ async def write_file(self, file_path: str, content: str, encoding: str = 'utf-8'
+                # mkdirs must run on the same sftp instance; if using ephemeral, do it inside the context
+                if not use_ephemeral:
+                    await _with_timeout(self._mkdirs(sftp, dirp), operation="create directories")
+            
+            # Write binary data directly
+            # Strategy 1: Write to a temporary file, verify size, then rename atomically
+            tmp_path = f"{path}.tmp.{secrets.token_hex(4)}"
+            try:
+                if use_ephemeral:
+                    async with hop.ephemeral_sftp() as esftp:
+                        if not esftp:
+                            raise RuntimeError("ephemeral SFTP unavailable")
```

Ruff (0.14.0)

Abstract `raise` to an inner function

(TRY301)